### PR TITLE
`Programming exercises`: Make hint display threshold editable

### PR DESCRIPTION
**Pull request by Hialus**
# Do not deploy this to a test server yet

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [ ] I implemented the changes with a good performance and prevented too many database calls.
- [ ] I documented the Java code using JavaDoc style.
#### Client
- [ ] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [ ] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
In the previous PR #5136, where we made the hint displaying dynamic, we introduced a fixed display threshold of 3. 
This PR allows editors to edit this threshold.

### Steps for Testing

Prerequisites:
- 1 Programming exercise with 1 existing hint
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to the programming exercise management page
3. Navigate to the hint management page for this exercise
4. Edit the existing hint
    - The display threshold should be set to 3
6. Create a new hint
    - The display threshold is set to 3 as a default
    - Set it to any value >= 0
8. Start the exercise as a student
    - No hints should be shown just yet
9. Push a submission
10. Any hints with a threshold of 0 are available
11. Push more submissions
12. Hints should be available after their threshold was met 

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|

### Screenshots

grafik

### CODE REVIEW

File: src/main/resources/config/liquibase/changelog/20220620150000_changelog.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.4.xsd">
+    <!--
+    This changeset adds the display threshold to the exercise_hint table
+    -->
+    <changeSet author="timor-morrien" id="20220620150000">
+        <addColumn tableName="exercise_hint">
+            <column name="display_threshold" type="INTEGER" defaultValueNumeric="3">
```
Comment (JohannesStoehr) [inline src/main/resources/config/liquibase/changelog/20220620150000_changelog.xml:10]: A smaller datatype, e.g. tinyint should be sufficient here. Additionally you can make this column unsigned

File: src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.ts [hunk 2]
```diff
@@ -63,6 +63,7 @@ export class ExerciseHintUpdateComponent implements OnInit, OnDestroy {
         });
         this.route.data.subscribe(({ exerciseHint }) => {
             this.exerciseHint = exerciseHint;
+            this.exerciseHint.displayThreshold = this.exerciseHint.displayThreshold ?? 3;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.ts:66]: You should put the 3 inside a constant somewhere

**Review by JohannesStoehr — DISMISSED** [review]
Left two comments

**Review by Marvin-Vitcu — DISMISSED** [review]
tested on ts1,
Tested with hint thresholds 0 and 5. Both worked fine.

File: src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.html [hunk 3]
```diff
@@ -32,6 +32,15 @@ <h2 id="jhi-exercise-hint-heading" [jhiTranslate]="'artemisApp.exerciseHint.home
                         <option *ngFor="let task of tasks" [ngValue]="task">{{ task.taskName }}</option>
                     </select>
                 </div>
+
+                <div class="form-group">
+                    <div>
+                        <label class="label-narrow" jhiTranslate="artemisApp.exerciseHint.displayThreshold" for="field_description">Display Threshold</label>
+                        <jhi-help-icon placement="auto" text="artemisApp.exerciseHint.displayThresholdTooltip"></jhi-help-icon>
+                    </div>
+                    <input type="number" required min="0" class="form-control" name="displayThreshold" id="field_displayThreshold" [(ngModel)]="exerciseHint.displayThreshold" />
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.html:41]: The threshold value is restricted to max. 100 on the server-side. Would it make sense to also force it here?

**Review by ole-ve — CHANGES_REQUESTED** [review]
Code looks added, I left some inline comments on some minors.
Also tested on TS5 with all edge cases I could imagine and everything worked as expected.

**Review by nairaarnez — CHANGES_REQUESTED** [review]
Tested in ts1, with numbers 0, 2, 3, 10, 20.Numbers >0 worked correctly. Hint with threshold 0 is not direct visible, it is shown after one submission 
image
image
In PR 5136 (https://github.com/ls1intum/Artemis/pull/5136)  the Student could also see the Hint in the main page of the Programming Exercice , now this was not visible 
image

File: src/main/webapp/i18n/de/exerciseHint.json [hunk 4]
```diff
@@ -20,6 +20,8 @@
             "title": "Titel",
             "task": "Task",
             "taskTooltip": "Es wird nicht empfohlen, Code-Hinweise anderen Tasks zuzuweisen, weil die Hinweise speziell für einen Task erstellt wurden.",
+            "displayThreshold": "Freischalten nach sovielen erfolglosen Einreichungen",
```
[the remark this answers is not in the document]
Comment (b-fein) [reply src/main/webapp/i18n/de/exerciseHint.json:23]:
```suggestion
            "displayThreshold": "Freischalten nach Anzahl erfolgloser Einreichungen",
```

File: src/main/webapp/i18n/en/exerciseHint.json [hunk 5]
```diff
@@ -20,6 +20,8 @@
             "title": "Title",
             "task": "Task",
             "taskTooltip": "It is not recommended to assign code hints to other tasks since these are specifically generated for a task.",
+            "displayThreshold": "Unlock after this many unsuccessful submissions",
```
[the remark this answers is not in the document]
Comment (b-fein) [reply src/main/webapp/i18n/en/exerciseHint.json:23]:
```suggestion
            "displayThreshold": "Unlock after number of unsuccessful submissions",
```

**Comment by ole-ve** [comment]
> In PR 5136 (https://github.com/ls1intum/Artemis/pull/5136) the Student could also see the Hint in the main page of the Programming Exercice , now this was not visible

Good catch, an analog corresponding change would be required in `course-exercise-details.component.ts`, l. 387ff.

**Comment by Hialus** [comment]
> Hint with threshold 0 is not direct visible, it is shown after one submission

This is the wanted behavior. Currently, the system only supports hints if there is at least one submission. So this is the expected behavior. 

Regarding your second point
> In PR 5136 (https://github.com/ls1intum/Artemis/pull/5136) the Student could also see the Hint in the main page of the Programming Exercice , now this was not visible

I amnot quite sure what you mean by that. Could you maybe elaborate a bit on this?

**Review by nadinesamyy — COMMENTED** [review]
tried testing on ts3, no programming exercises were displayed even in courses that should have programming exercises and no new ones could be imported - caused an internal server error

**Comment by Hialus** [comment]
> Thank you for the clarification! So Hints with threshold 0 and 1 have the same behavior?

They are somewhat similar at the moment, but hints with a threshold 1 will only be displayed if they are assigned to the next unsuccessful task. Hints with a threshold of 0 will be displayed regardless of the status of any task.

**Review by Dominik-Weinzierl — APPROVED** [review]
Tested during the testing session - works as expected

**Review by ge65cer — COMMENTED** [review]
Functionality works as expected.
Server tests fail, feel free to ping me after they run through. 🙂

**Review by ge65cer — APPROVED** [review]
LGTM

**Review by Dominik-Weinzierl — APPROVED** [review]
Looks good to me: The issues regarding the failing tests and the formatting are solved in https://github.com/ls1intum/Artemis/pull/5391 (so not introduced in this PR)
